### PR TITLE
SW-1842 Only do fuzzy matching if there are no exact matches

### PR DESCRIPTION
**Pull request by sgrimm**
We want users to be able to type nonexistent accession numbers into the accession
list UI and get back close matches. But we also want users to be able to paste
exact accession numbers into the same search field and not have the results
cluttered with accessions they didn't search for.

Try to thread this needle by making the fuzzy search code start off by trying an
exact search. If that finds matches, then we're probably in the "user entered an
exact value" situation; return those results. Otherwise do the fuzzy search and
return approximate matches.

This change means there is no way for a user to do a fuzzy search for a value that
also happens to be an exact match; we believe that will still be an improvement
over the current behavior.

### CODE REVIEW

**Comment by sgrimm** [comment]
Current dependencies on/for this PR:
* main
  * **PR #764** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/764" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/764?utm_source=stack-comment).

File: src/main/kotlin/com/terraformation/backend/search/SearchFilter.kt [hunk 1]
```diff
@@ -12,15 +12,33 @@ enum class SearchFilterType {
...
+   * currently contain any fuzzy filter criteria, this method _must_ return an object that tests
+   * equal to `this` (e.g., by just returning `this`).
+   */
+  fun toExactSearch(): SearchNode
+
   fun referencedSublists(): Set<SublistField>
 }
 
-class OrNode(private val children: List<SearchNode>) : SearchNode {
+data class OrNode(private val children: List<SearchNode>) : SearchNode {
```
Comment (charlie-tf) [inline src/main/kotlin/com/terraformation/backend/search/SearchFilter.kt:26]: nice usage of data classes. Is a one of the benefits here that it provides equality checking?

File: src/main/kotlin/com/terraformation/backend/search/SearchFilter.kt [hunk 2]
```diff
@@ -59,7 +97,7 @@ class NotNode(val child: SearchNode) : SearchNode {
   }
 }
 
-class FieldNode(
+data class FieldNode(
```
Comment (charlie-tf) [inline src/main/kotlin/com/terraformation/backend/search/SearchFilter.kt:100]: Are `FieldNode`s the leaf nodes in the search?

File: src/test/kotlin/com/terraformation/backend/seedbank/search/SearchServiceFuzzySearchTest.kt [hunk 3]
```diff
@@ -43,4 +43,20 @@ internal class SearchServiceFuzzySearchTest : SearchServiceTest() {
...
     assertEquals(expected, result)
   }
+
+  @Test
+  fun `fuzzy search on text fields tries exact matching first`() {
+    accessionsDao.update(accessionsDao.fetchOneById(AccessionId(1000))!!.copy(number = "22-1-100"))
+    accessionsDao.update(accessionsDao.fetchOneById(AccessionId(1001))!!.copy(number = "22-1-101"))
+
+    val fields = listOf(accessionNumberField)
+    val searchNode = FieldNode(accessionNumberField, listOf("22-1-100"), SearchFilterType.Fuzzy)
```
Comment (charlie-tf) [inline src/test/kotlin/com/terraformation/backend/seedbank/search/SearchServiceFuzzySearchTest.kt:53]: in this case, would a fuzzy search also return `22-1-101`?

**Review by charlie-tf — APPROVED** [review]
looks good! just a few questions to check my understanding

File: src/main/kotlin/com/terraformation/backend/search/SearchFilter.kt (hunk 1 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/search/SearchFilter.kt:26]: Yes, I did this to get the `equals` implementation.

File: src/main/kotlin/com/terraformation/backend/search/SearchFilter.kt (hunk 2 above)
Comment (sgrimm) [reply src/main/kotlin/com/terraformation/backend/search/SearchFilter.kt:100]: > Are `FieldNode`s the leaf nodes in the search?

Yep!

File: src/test/kotlin/com/terraformation/backend/seedbank/search/SearchServiceFuzzySearchTest.kt (hunk 3 above)
Comment (sgrimm) [reply src/test/kotlin/com/terraformation/backend/seedbank/search/SearchServiceFuzzySearchTest.kt:53]: > in this case, would a fuzzy search also return `22-1-101`?

Yes. I'll update the test to make that more explicit.

**Review by charlie-tf — APPROVED** [review]

**Pull request merged**